votor/src/consensus_pool/certificate_builder.rs

```diff
@@ -16,87 +16,288 @@ use {
+        match self {
+            Self::SingleVote { signature, bitmap } => {
+                build_cert_from_bitmap(certificate, signature, bitmap).map_err(BuildError::Encode)
             }
-            Some(mut bitmap_1) => {
-                let last_one_0 = self.bitmap_0.last_one().map_or(0, |i| i.saturating_add(1));
-                let last_one_1 = bitmap_1.last_one().map_or(0, |i| i.saturating_add(1));
-                let new_length = last_one_0.max(last_one_1);
-                self.bitmap_0.resize(new_length, false);
-                bitmap_1.resize(new_length, false);
-                encode_base3(&self.bitmap_0, &bitmap_1).map_err(CertificateError::EncodeError)?
-            }
-        };
-        Ok(CertificateMessage {
-            certificate: self.certificate,
-            signature: self.signature.into(),
-            bitmap,
-        })
+            Self::Skip {
+                mut signature0,
+                bitmap0,
+                sig_and_bitmap1,
+            } => match sig_and_bitmap1 {
+                None => build_cert_from_bitmap(certificate, signature0, bitmap0)
+                    .map_err(BuildError::Encode),
+                Some((signature1, bitmap1)) => {
+                    signature0.aggregate_with([signature1].iter())?;
+                    build_cert_from_bitmaps(certificate, signature0, bitmap0, bitmap1)
+                }
+            },
+            Self::NotarFallback {
+                signature,
+                bitmap0,
+                bitmap1,
+            } => match bitmap1 {
+                None => build_cert_from_bitmap(certificate, signature, bitmap0)
+                    .map_err(BuildError::Encode),
+                Some(bitmap1) => build_cert_from_bitmaps(certificate, signature, bitmap0, bitmap1),
+            },
+        }
+    }
+
+    /// Builds a [`CertificateMessage`] for rewards purposes from the builder.
+    fn build_for_rewards(
+        self,
+        certificate: Certificate,
+    ) -> Result<CertificateMessage, BuildForRewardsError> {
+        match self {
+            Self::Skip {
+                signature0,
+                bitmap0,
+                sig_and_bitmap1: _,
+            } => build_cert_from_bitmap(certificate, signature0, bitmap0)
+                .map_err(BuildForRewardsError::Encode),
+            Self::SingleVote { signature, bitmap } => match certificate {
+                Certificate::Notarize(_, _) => {
+                    build_cert_from_bitmap(certificate, signature, bitmap)
+                        .map_err(BuildForRewardsError::Encode)
+                }
+                _ => Err(BuildForRewardsError::InvalidCertType),
+            },
+            Self::NotarFallback { .. } => Err(BuildForRewardsError::InvalidCertType),
+        }
+    }
+}
+
+/// Builder for creating [`CertificateMessage`] by using BLS signature aggregation.
+pub struct CertificateBuilder {
+    builder_type: BuilderType,
+    certificate: Certificate,
```

Comment:
not part of this PR, but the type name `Certificate` is *very* confusing to me, because it does *not* include the aggregate signature, right?